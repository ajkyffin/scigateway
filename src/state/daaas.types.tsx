import { Step } from 'react-joyride';

export const AuthSuccessType = 'daaas:auth_success';
export const AuthFailureType = 'daaas:auth_failure';
export const LoadingAuthType = 'daaas:loading_auth';
export const LoadAuthProviderType = 'daaas:loading_auth_provider';
export const TokenExpiredType = 'daaas:expired_token';
export const ConfigureStringsType = 'daaas:configure_strings';
export const LoginType = 'daaas:login';
export const NotificationType = 'daaas:api:notification';
export const RegisterRouteType = 'daaas:api:register_route';
export const RequestPluginRerenderType = 'daaas:api:plugin_rerender';
export const SignOutType = 'daaas:signout';
export const ToggleDrawerType = 'daaas:toggledrawer';
export const DismissNotificationType = 'daaas:dismissnotification';
export const ConfigureFeatureSwitchesType = 'daaas:feature_switches';
export const AddNotificationsType = 'daaas:add_notification';
export const SiteLoadingType = 'daaas:site_loading';
export const ConfigureAnalyticsType = 'daaas:configure_analytics';
export const InitialiseAnalyticsType = 'daaas:initialise_analytics';
export const ToggleHelpType = 'daaas:toggle_help';
export const AddHelpTourStepsType = 'daaas:add_help_tour_steps';

export interface NotificationPayload {
  message: string;
  severity: string;
}

export interface AddNotificationsPayload {
  notifications: NotificationPayload[];
}

export interface AppStrings {
  [id: string]: string;
}

export interface ConfigureStringsPayload {
  res: ApplicationStrings;
}

export interface ApplicationStrings {
  [section: string]: AppStrings;
}

export interface FeatureSwitchesPayload {
  switches: FeatureSwitches;
}

export interface FeatureSwitches {
  showContactButton: boolean;
}

export interface RegisterRoutePayload {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
}

export interface PluginConfig {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
}

export interface GroupedPlugins {
  [section: string]: PluginConfig[];
}

export interface LoginPayload {
  username: string;
  password: string;
}

export interface AccessPayload {
  authorisation: boolean;
  accessToken: string;
  error: null;
}

export interface AuthorisedPayload {
  token: string;
}

export interface DismissNotificationPayload {
  index: number;
}

export interface AuthProviderPayload {
  authProvider: string;
}

export interface SiteLoadingPayload {
  loading: boolean;
}

export interface ConfigureAnalyticsPayload {
  id: string;
}

export interface AddHelpTourStepsPayload {
  steps: Step[];
}
