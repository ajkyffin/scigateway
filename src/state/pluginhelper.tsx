import {
  GroupedPlugins,
  PluginConfig,
  RegisterRoutePayload,
} from './daaas.types';

// Custom comparer for plugin objects, sorts keys by order (asc), then by name (asc)
export const comparePlugins = (a: PluginConfig, b: PluginConfig): number =>
  a.order === b.order
    ? a.displayName.localeCompare(b.displayName)
    : a.order - b.order;

function buildMenuItemDictionary(plugins: PluginConfig[]): GroupedPlugins {
  const dict: GroupedPlugins = {};

  plugins.forEach(p => {
    if (!(p.section in dict)) {
      dict[p.section] = [];
    }
    dict[p.section].push(p);
  });

  return dict;
}

// Convert the list of plugins into a structured dataset for rendering
export function structureMenuData(plugins: PluginConfig[]): GroupedPlugins {
  const menuItems = buildMenuItemDictionary(plugins);

  Object.keys(menuItems).forEach(
    section => (menuItems[section] = menuItems[section].sort(comparePlugins))
  );
  return menuItems;
}

// Convert a plugin registration request body into a pluginConfig object
export function buildPluginConfig(payload: RegisterRoutePayload): PluginConfig {
  return { ...payload };
}
