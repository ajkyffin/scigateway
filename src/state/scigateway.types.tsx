import { Step } from 'react-joyride';

export const AuthSuccessType = 'scigateway:auth_success';
export const AuthFailureType = 'scigateway:auth_failure';
export const LoadingAuthType = 'scigateway:loading_auth';
export const LoadAuthProviderType = 'scigateway:loading_auth_provider';
export const LoadAuthUrlType = 'scigateway:loading_auth_url';
export const InvalidateTokenType = 'scigateway:api:invalidate_token';
export const ConfigureStringsType = 'scigateway:configure_strings';
export const LoginType = 'scigateway:login';
export const NotificationType = 'scigateway:api:notification';
export const RegisterRouteType = 'scigateway:api:register_route';
export const RequestPluginRerenderType = 'scigateway:api:plugin_rerender';
export const SignOutType = 'scigateway:signout';
export const ToggleDrawerType = 'scigateway:toggledrawer';
export const DismissNotificationType = 'scigateway:dismissnotification';
export const ConfigureFeatureSwitchesType = 'scigateway:feature_switches';
export const AddNotificationsType = 'scigateway:add_notification';
export const SiteLoadingType = 'scigateway:site_loading';
export const ConfigureAnalyticsType = 'scigateway:configure_analytics';
export const InitialiseAnalyticsType = 'scigateway:initialise_analytics';
export const ToggleHelpType = 'scigateway:toggle_help';
export const AddHelpTourStepsType = 'scigateway:add_help_tour_steps';

export interface NotificationPayload {
  message: string;
  severity: string;
}

export interface AddNotificationsPayload {
  notifications: NotificationPayload[];
}

export interface AppStrings {
  [id: string]: string;
}

export interface ConfigureStringsPayload {
  res: ApplicationStrings;
}

export interface ApplicationStrings {
  [section: string]: AppStrings;
}

export interface FeatureSwitchesPayload {
  switches: FeatureSwitches;
}

export interface FeatureSwitches {
  showContactButton: boolean;
}

export interface RegisterRoutePayload {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
}

export interface PluginConfig {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
}

export interface GroupedPlugins {
  [section: string]: PluginConfig[];
}

export interface LoginPayload {
  username: string;
  password: string;
}

export interface AccessPayload {
  authorisation: boolean;
  accessToken: string;
  error: null;
}

export interface AuthorisedPayload {
  token: string;
}

export interface DismissNotificationPayload {
  index: number;
}

export interface AuthProviderPayload {
  authProvider: string;
  authUrl: string | undefined;
}

export interface SiteLoadingPayload {
  loading: boolean;
}

export interface ConfigureAnalyticsPayload {
  id: string;
}

export interface AddHelpTourStepsPayload {
  steps: Step[];
}
