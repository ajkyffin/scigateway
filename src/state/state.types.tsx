import { ThunkAction } from 'redux-thunk';
import { AnyAction } from 'redux';
import { Step } from 'react-joyride';
import {
  ApplicationStrings,
  PluginConfig,
  FeatureSwitches,
} from './scigateway.types';
import { RouterState } from 'connected-react-router';

export interface Plugin {
  name: string;
  src: string;
  enable: boolean;
  location: 'main' | 'left' | 'right';
}

export interface ScigatewayNotification {
  message: string;
  severity: string;
}

export interface ScigatewayState {
  notifications: ScigatewayNotification[];
  plugins: PluginConfig[];
  drawerOpen: boolean;
  siteLoading: boolean;
  showHelp: boolean;
  helpSteps: Step[];
  authorisation: AuthState;
  res?: ApplicationStrings;
  features: FeatureSwitches;
  analytics?: AnalyticsState;
}

export interface StateType {
  scigateway: ScigatewayState;
  router: RouterState;
}

export interface ActionType<T> {
  type: string;
  payload: T;
}

export type ThunkResult<R> = ThunkAction<R, StateType, null, AnyAction>;

export interface User {
  username: string;
  avatarUrl: string;
}

// eslint-disable-next-line @typescript-eslint/interface-name-prefix
export interface ICATAuthenticator {
  mnemonic: string;
  keys: { name: string; hide?: boolean }[];
  friendly?: string;
  admin?: boolean;
}

export interface AuthProvider {
  isLoggedIn: () => boolean;
  logOut: () => void;
  logIn: (username: string, password: string) => Promise<void>;
  verifyLogIn: () => Promise<void>;
  refresh: () => Promise<void>;
  redirectUrl: string | null;
  user: User | null;
  mnemonic?: string;
  autoLogin?: () => Promise<void>;
}

export interface AuthState {
  failedToLogin: boolean;
  signedOutDueToTokenInvalidation: boolean;
  loading: boolean;
  provider: AuthProvider;
}

export interface AnalyticsState {
  id: string;
  initialised: boolean;
}
