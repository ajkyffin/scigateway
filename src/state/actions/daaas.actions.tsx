import { NotificationType, NotificationPayload } from '../daaas.types';
import { ActionType, StateType } from '../state.types';
import axios from 'axios';
import { ThunkAction } from 'redux-thunk';
import { AnyAction } from 'redux';

export const daaasNotification = (
  message: string
): ActionType<NotificationPayload> => ({
  type: NotificationType,
  payload: {
    message,
  },
});

type MyThunkResult<R> = ThunkAction<R, StateType, null, AnyAction>;

export const configureSite = (): MyThunkResult<Promise<void>> => {
  return async dispatch => {
    const res = await axios.get(`/settings.json`);
    const settings = res.data;
    dispatch(daaasNotification(JSON.stringify(settings)));
  };
};
