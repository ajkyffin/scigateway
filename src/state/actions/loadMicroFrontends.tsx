/* eslint-disable @typescript-eslint/explicit-function-return-type */
// single-spa doesn't come with any types - all single-spa code should be limited to this file.
import * as singleSpa from 'single-spa';
import { Plugin } from '../state.types';

const runScript = async (url: string) => {
  return new Promise((resolve, reject) => {
    const script = document.createElement('script');
    script.src = url;
    script.onload = resolve;
    script.onerror = reject;

    const body = document.getElementsByTagName('body')[0];
    body.appendChild(script);
  });
};

const loadReactApp = async (name: string, url: string) => {
  await runScript(url);
  // Plugins are loaded on to the window and define their own property name so can't guess this at compile time
  /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
  return (window as any)[name];
};

async function loadApp(name: string, appURL: string) {
  // register the app with singleSPA and pass a reference to the store of the app as well as a reference to the globalEventDistributor
  singleSpa.registerApplication(
    name,
    () => loadReactApp(name, appURL),
    (location: URL) => {
      return location.pathname === '/';
    }
  );
}

async function init(plugins: Plugin[]) {
  /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
  const loadingPromises: Promise<any>[] = [];

  plugins
    .filter(p => p.enable)
    .forEach(p => {
      loadingPromises.push(loadApp(p.name, p.src));
    });

  // wait until all stores are loaded and all apps are registered with singleSpa
  await Promise.all(loadingPromises);

  singleSpa.start();
}

export default {
  init,
};
