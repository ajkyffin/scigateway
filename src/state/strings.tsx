import { AppStrings } from './scigateway.types';
import { StateType } from './state.types';

export function getAppStrings(
  state: StateType,
  section: string
): AppStrings | undefined {
  return state.scigateway.res ? state.scigateway.res[section] : undefined;
}

export const getString = (res: AppStrings | undefined, key: string): string =>
  (res && res[key]) || key;
