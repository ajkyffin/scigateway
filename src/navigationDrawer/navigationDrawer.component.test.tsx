import React from 'react';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import { StateType } from '../state/state.types';
import configureStore from 'redux-mock-store';
import { initialState } from '../state/reducers/daaas.reducer';
import NavigationDrawer from './navigationDrawer.component';
import { toggleDrawer } from '../state/actions/daaas.actions';

describe('Main app bar component', () => {
  let shallow;
  let mount;
  let mockStore;
  let state: StateType;

  beforeEach(() => {
    shallow = createShallow({});
    mount = createMount();

    mockStore = configureStore();
    state = {
      daaas: initialState,
    };
  });

  afterEach(() => {
    mount.cleanUp();
  });

  it('Navigation drawer renders correctly when open', () => {
    state.daaas.drawerOpen = true;

    const wrapper = shallow(<NavigationDrawer store={mockStore(state)} />);
    expect(wrapper.dive().dive()).toMatchSnapshot();
  });

  it('Navigation drawer renders correctly when closed', () => {
    state.daaas.drawerOpen = false;

    const wrapper = shallow(<NavigationDrawer store={mockStore(state)} />);
    expect(wrapper.dive().dive()).toMatchSnapshot();
  });

  it('sends toggleDrawer action when chevron clicked', () => {
    const testStore = mockStore(state);
    const wrapper = mount(<NavigationDrawer store={testStore} />);

    wrapper
      .find('button')
      .first()
      .simulate('click');

    expect(testStore.getActions().length).toEqual(1);
    expect(testStore.getActions()[0]).toEqual(toggleDrawer());
  });
});
