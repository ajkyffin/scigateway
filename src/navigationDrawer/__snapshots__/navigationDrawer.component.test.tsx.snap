// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main app bar component Navigation drawer renders correctly when closed 1`] = `
<WithStyles(Drawer)
  anchor="left"
  className="NavigationDrawer-drawer-1"
  classes={
    Object {
      "paper": "NavigationDrawer-drawerPaper-2",
    }
  }
  open={false}
  variant="persistent"
>
  <div
    className="NavigationDrawer-drawerHeader-3"
  >
    <WithStyles(IconButton)
      onClick={[Function]}
    >
      <pure(ChevronLeftIcon) />
    </WithStyles(IconButton)>
  </div>
  <WithStyles(Divider) />
  <WithStyles(List) />
</WithStyles(Drawer)>
`;

exports[`Main app bar component Navigation drawer renders correctly when open 1`] = `
<WithStyles(Drawer)
  anchor="left"
  className="NavigationDrawer-drawer-1"
  classes={
    Object {
      "paper": "NavigationDrawer-drawerPaper-2",
    }
  }
  open={true}
  variant="persistent"
>
  <div
    className="NavigationDrawer-drawerHeader-3"
  >
    <WithStyles(IconButton)
      onClick={[Function]}
    >
      <pure(ChevronLeftIcon) />
    </WithStyles(IconButton)>
  </div>
  <WithStyles(Divider) />
  <WithStyles(List) />
</WithStyles(Drawer)>
`;

exports[`Main app bar component renders a plugin list grouped by sections ordered alphabetically when open 1`] = `
<WithStyles(Drawer)
  anchor="left"
  className="NavigationDrawer-drawer-1"
  classes={
    Object {
      "paper": "NavigationDrawer-drawerPaper-2",
    }
  }
  open={true}
  variant="persistent"
>
  <div
    className="NavigationDrawer-drawerHeader-3"
  >
    <WithStyles(IconButton)
      onClick={[Function]}
    >
      <pure(ChevronLeftIcon) />
    </WithStyles(IconButton)>
  </div>
  <WithStyles(Divider) />
  <WithStyles(List)>
    <WithStyles(ListItem)>
      <WithStyles(ListItemText)
        classes={
          Object {
            "primary": "NavigationDrawer-sectionTitle-4",
          }
        }
        primary="ANALYSIS"
        primaryTypographyProps={
          Object {
            "variant": "h6",
          }
        }
      />
    </WithStyles(ListItem)>
    <WithStyles(List)
      component="nav"
    >
      <LinkListItem
        key="0"
        to="link"
      >
        <WithStyles(ListItemText)
          classes={
            Object {
              "primary": "NavigationDrawer-menuItem-5",
            }
          }
          inset={true}
          primary="analysis-plugin"
          primaryTypographyProps={
            Object {
              "variant": "subtitle1",
            }
          }
        />
      </LinkListItem>
      <LinkListItem
        key="1"
        to="link"
      >
        <WithStyles(ListItemText)
          classes={
            Object {
              "primary": "NavigationDrawer-menuItem-5",
            }
          }
          inset={true}
          primary="analysis-plugin2"
          primaryTypographyProps={
            Object {
              "variant": "subtitle1",
            }
          }
        />
      </LinkListItem>
    </WithStyles(List)>
    <WithStyles(ListItem)>
      <WithStyles(ListItemText)
        classes={
          Object {
            "primary": "NavigationDrawer-sectionTitle-4",
          }
        }
        primary="DATA"
        primaryTypographyProps={
          Object {
            "variant": "h6",
          }
        }
      />
    </WithStyles(ListItem)>
    <WithStyles(List)
      component="nav"
    >
      <LinkListItem
        key="0"
        to="link"
      >
        <WithStyles(ListItemText)
          classes={
            Object {
              "primary": "NavigationDrawer-menuItem-5",
            }
          }
          inset={true}
          primary="data-plugin"
          primaryTypographyProps={
            Object {
              "variant": "subtitle1",
            }
          }
        />
      </LinkListItem>
      <LinkListItem
        key="1"
        to="link"
      >
        <WithStyles(ListItemText)
          classes={
            Object {
              "primary": "NavigationDrawer-menuItem-5",
            }
          }
          inset={true}
          primary="data-plugin1"
          primaryTypographyProps={
            Object {
              "variant": "subtitle1",
            }
          }
        />
      </LinkListItem>
      <LinkListItem
        key="2"
        to="link"
      >
        <WithStyles(ListItemText)
          classes={
            Object {
              "primary": "NavigationDrawer-menuItem-5",
            }
          }
          inset={true}
          primary="data-plugin2"
          primaryTypographyProps={
            Object {
              "variant": "subtitle1",
            }
          }
        />
      </LinkListItem>
      <LinkListItem
        key="3"
        to="plugin_link"
      >
        <WithStyles(ListItemText)
          classes={
            Object {
              "primary": "NavigationDrawer-menuItem-5",
            }
          }
          inset={true}
          primary="data-plugin-no-displayname"
          primaryTypographyProps={
            Object {
              "variant": "subtitle1",
            }
          }
        />
      </LinkListItem>
    </WithStyles(List)>
  </WithStyles(List)>
</WithStyles(Drawer)>
`;

exports[`Main app bar component renders a plugin list grouped by sections ordered alphabetically when open 2`] = `
<WithStyles(ListItem)
  button={true}
  component={[Function]}
  to="plugin_link"
>
  <WithStyles(ListItemText)
    classes={
      Object {
        "primary": "NavigationDrawer-menuItem-5",
      }
    }
    inset={true}
    primary="data-plugin-no-displayname"
    primaryTypographyProps={
      Object {
        "variant": "subtitle1",
      }
    }
  />
</WithStyles(ListItem)>
`;
